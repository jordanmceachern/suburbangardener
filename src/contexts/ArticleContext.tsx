"use client";

import React, {
  createContext,
  useContext,
  useReducer,
  useCallback,
} from "react";
import type { ArticleDisplayData } from "../components/articles/ArticleDisplayCard";

interface ArticleState {
  articles: ArticleDisplayData[];
  featuredArticle: ArticleDisplayData | null;
  total: number;
  hasMore: boolean;
  loading: boolean;
  error: string | null;
  searchTerm: string;
  selectedCategory: string;
  currentOffset: number;
}

type ArticleAction =
  | { type: "SET_LOADING"; payload: boolean }
  | { type: "SET_ERROR"; payload: string | null }
  | {
      type: "SET_ARTICLES";
      payload: {
        articles: ArticleDisplayData[];
        total: number;
        hasMore: boolean;
      };
    }
  | {
      type: "APPEND_ARTICLES";
      payload: { articles: ArticleDisplayData[]; hasMore: boolean };
    }
  | { type: "SET_FEATURED_ARTICLE"; payload: ArticleDisplayData | null }
  | { type: "SET_SEARCH_TERM"; payload: string }
  | { type: "SET_SELECTED_CATEGORY"; payload: string }
  | { type: "SET_OFFSET"; payload: number }
  | { type: "RESET_PAGINATION" };

const initialState: ArticleState = {
  articles: [],
  featuredArticle: null,
  total: 0,
  hasMore: false,
  loading: false,
  error: null,
  searchTerm: "",
  selectedCategory: "All Categories",
  currentOffset: 0,
};

function articleReducer(
  state: ArticleState,
  action: ArticleAction
): ArticleState {
  switch (action.type) {
    case "SET_LOADING":
      return { ...state, loading: action.payload };
    case "SET_ERROR":
      return { ...state, error: action.payload, loading: false };
    case "SET_ARTICLES":
      return {
        ...state,
        articles: action.payload.articles,
        total: action.payload.total,
        hasMore: action.payload.hasMore,
        loading: false,
        error: null,
        currentOffset: action.payload.articles.length,
      };
    case "APPEND_ARTICLES":
      return {
        ...state,
        articles: [...state.articles, ...action.payload.articles],
        hasMore: action.payload.hasMore,
        loading: false,
        currentOffset: state.currentOffset + action.payload.articles.length,
      };
    case "SET_FEATURED_ARTICLE":
      return { ...state, featuredArticle: action.payload };
    case "SET_SEARCH_TERM":
      return { ...state, searchTerm: action.payload };
    case "SET_SELECTED_CATEGORY":
      return { ...state, selectedCategory: action.payload };
    case "SET_OFFSET":
      return { ...state, currentOffset: action.payload };
    case "RESET_PAGINATION":
      return { ...state, currentOffset: 0, hasMore: false };
    default:
      return state;
  }
}

interface ArticleContextType {
  state: ArticleState;
  loadArticles: () => Promise<void>;
  loadMoreArticles: () => Promise<void>;
  searchArticles: (searchTerm: string) => Promise<void>;
  setSelectedCategory: (category: string) => void;
  getFilteredArticles: () => ArticleDisplayData[];
}

const ArticleContext = createContext<ArticleContextType | null>(null);

export function useArticles() {
  const context = useContext(ArticleContext);
  if (!context) {
    throw new Error("useArticles must be used within an ArticleProvider");
  }
  return context;
}

interface ArticleProviderProps {
  children: React.ReactNode;
}

export function ArticleProvider({ children }: ArticleProviderProps) {
  const [state, dispatch] = useReducer(articleReducer, initialState);

  const loadArticles = useCallback(
    async (reset = false) => {
      dispatch({ type: "SET_LOADING", payload: true });
      dispatch({ type: "SET_ERROR", payload: null });

      if (reset) {
        dispatch({ type: "RESET_PAGINATION" });
      }

      try {
        const offset = reset ? 0 : state.currentOffset;
        const params = new URLSearchParams({
          limit: "20",
          offset: offset.toString(),
          ...(state.searchTerm && { search: state.searchTerm }),
        });

        const response = await fetch(`/api/articles?${params}`);
        if (!response.ok) {
          throw new Error("Failed to fetch articles");
        }

        const data = await response.json();

        if (reset || offset === 0) {
          dispatch({
            type: "SET_ARTICLES",
            payload: {
              articles: data.articles,
              total: data.total,
              hasMore: data.hasMore,
            },
          });

          // Set featured article (first article)
          if (data.articles.length > 0) {
            dispatch({
              type: "SET_FEATURED_ARTICLE",
              payload: data.articles[0],
            });
          }
        } else {
          dispatch({
            type: "APPEND_ARTICLES",
            payload: {
              articles: data.articles,
              hasMore: data.hasMore,
            },
          });
        }
      } catch (error) {
        dispatch({
          type: "SET_ERROR",
          payload: error instanceof Error ? error.message : "An error occurred",
        });
      }
    },
    [state.currentOffset, state.searchTerm]
  );

  const loadMoreArticles = useCallback(async () => {
    if (!state.hasMore || state.loading) return;
    await loadArticles(false);
  }, [state.hasMore, state.loading, loadArticles]);

  const searchArticles = useCallback(
    async (searchTerm: string) => {
      dispatch({ type: "SET_SEARCH_TERM", payload: searchTerm });
      await loadArticles(true);
    },
    [loadArticles]
  );

  const setSelectedCategory = useCallback((category: string) => {
    dispatch({ type: "SET_SELECTED_CATEGORY", payload: category });
  }, []);

  const getFilteredArticles = useCallback(() => {
    let filtered = state.articles;

    // Exclude featured article from the list to avoid duplication
    if (state.featuredArticle) {
      filtered = filtered.filter(
        article => article.id !== state.featuredArticle!.id
      );
    }

    // Apply category filter (client-side filtering)
    if (state.selectedCategory !== "All Categories") {
      filtered = filtered.filter(
        article => article.category === state.selectedCategory
      );
    }

    return filtered;
  }, [state.articles, state.selectedCategory, state.featuredArticle]);

  const contextValue: ArticleContextType = {
    state,
    loadArticles,
    loadMoreArticles,
    searchArticles,
    setSelectedCategory,
    getFilteredArticles,
  };

  return (
    <ArticleContext.Provider value={contextValue}>
      {children}
    </ArticleContext.Provider>
  );
}
