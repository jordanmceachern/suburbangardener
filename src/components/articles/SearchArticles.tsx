"use client";

import React, { useState, useEffect, useRef } from "react";
import { useArticles } from "../../contexts/ArticleContext";

interface SearchArticlesProps {
  className?: string;
}

export default function SearchArticles({
  className = "",
}: SearchArticlesProps) {
  const { state, searchArticles, setSelectedCategory } = useArticles();
  const [localSearchTerm, setLocalSearchTerm] = useState("");
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  // Get unique categories from articles
  const categories = [
    "All Categories",
    ...Array.from(new Set(state.articles.map(article => article.category))),
  ];

  // Close dropdown when clicking outside
  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsDropdownOpen(false);
      }
    }

    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  const handleSearch = async () => {
    if (
      localSearchTerm.trim().length >= 4 ||
      localSearchTerm.trim().length === 0
    ) {
      await searchArticles(localSearchTerm.trim());
    }
  };

  const handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === "Enter") {
      handleSearch();
    }
  };

  const handleCategoryChange = (category: string) => {
    setSelectedCategory(category);
    setIsDropdownOpen(false);
  };

  const filteredArticles = state.articles.filter(article => {
    if (state.selectedCategory === "All Categories") return true;
    return article.category === state.selectedCategory;
  });

  const toggleCategoryFilter = () => setIsDropdownOpen(!isDropdownOpen);

  return (
    <div className={`${className}`}>
      <h3 className="text-lg font-semibold text-slate-800 dark:text-slate-200 mb-4">
        Search Articles
      </h3>

      {/* Search and Filter Bar */}
      <div className="space-y-4">
        {/* Search Input */}
        <div className="relative">
          <input
            type="text"
            placeholder="Search articles..."
            value={localSearchTerm}
            onChange={e => setLocalSearchTerm(e.target.value)}
            onKeyPress={handleKeyPress}
            className="block w-full pl-3 pr-12 py-2.5 h-10 text-base border border-slate-300 dark:border-slate-600 rounded-md bg-white dark:bg-slate-700 text-slate-900 dark:text-slate-100 placeholder-slate-900 dark:placeholder-slate-100 focus:outline-none focus:ring-2 focus:ring-primary-500 focus:border-primary-500 cursor-text hover:border-slate-400 dark:hover:border-slate-500 transition-colors"
          />
          <button
            onClick={handleSearch}
            disabled={
              localSearchTerm.trim().length > 0 &&
              localSearchTerm.trim().length < 4
            }
            className="absolute inset-y-0 right-0 pr-3 flex items-center text-slate-900 dark:text-slate-100 hover:text-slate-600 dark:hover:text-slate-300 disabled:opacity-50 disabled:cursor-not-allowed"
          >
            <svg
              className="h-4 w-4"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
              />
            </svg>
          </button>
        </div>

        {/* Category Filter */}
        <div ref={dropdownRef} className="relative">
          <button
            type="button"
            onClick={toggleCategoryFilter}
            className="w-full px-3 py-2.5 h-10 text-base border border-slate-300 dark:border-slate-600 rounded-md bg-white dark:bg-slate-700 text-slate-900 dark:text-slate-100 focus:outline-none focus:ring-2 focus:ring-primary-500 focus:border-primary-500 cursor-pointer hover:border-slate-400 dark:hover:border-slate-500 transition-colors text-left flex items-center justify-between"
            aria-haspopup="listbox"
            aria-expanded={isDropdownOpen}
            aria-label="Select category"
          >
            <span className="truncate">{state.selectedCategory}</span>
            <svg
              className={`h-4 w-4 text-slate-900 dark:text-slate-100 transition-transform ${
                isDropdownOpen ? "transform rotate-180" : ""
              }`}
              fill="none"
              stroke="currentColor"
              viewBox="0 0 20 20"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={1.5}
                d="M6 8l4 4 4-4"
              />
            </svg>
          </button>

          {isDropdownOpen && (
            <div className="absolute z-50 w-full mt-1 bg-white dark:bg-slate-700 border border-slate-300 dark:border-slate-600 rounded-md max-h-60 overflow-auto shadow-lg">
              {categories.map(category => (
                <button
                  key={category}
                  type="button"
                  onClick={() => handleCategoryChange(category)}
                  className={`w-full px-3 py-2.5 text-left text-base hover:bg-slate-50 dark:hover:bg-slate-600 focus:outline-none focus:bg-slate-50 dark:focus:bg-slate-600 transition-colors first:rounded-t-md last:rounded-b-md ${
                    state.selectedCategory === category
                      ? "bg-primary-50 dark:bg-primary-900/20 text-primary-700 dark:text-primary-300"
                      : "text-slate-900 dark:text-slate-100"
                  }`}
                  role="option"
                  aria-selected={state.selectedCategory === category}
                >
                  {category}
                </button>
              ))}
            </div>
          )}
        </div>

        {/* Results Summary */}
        <div className="text-xs text-slate-600 dark:text-slate-300">
          {state.searchTerm || state.selectedCategory !== "All Categories" ? (
            <span>
              Found {filteredArticles.length} articles
              {state.searchTerm && ` for "${state.searchTerm}"`}
            </span>
          ) : (
            <span>{state.total} total articles</span>
          )}
        </div>
      </div>
    </div>
  );
}
