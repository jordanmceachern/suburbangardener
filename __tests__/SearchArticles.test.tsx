import { render } from "@testing-library/react";
import SearchArticles from "@/components/articles/SearchArticles";
import { ArticleProvider } from "@/contexts/ArticleContext";

describe("SearchArticles Component", () => {
  const mockProps = {
    className: "test-search-class",
  };

  it("should match DOM snapshot", () => {
    const { container } = render(
      <ArticleProvider>
        <SearchArticles {...mockProps} />
      </ArticleProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should maintain CSS classes for search input", () => {
    const { container } = render(
      <ArticleProvider>
        <SearchArticles {...mockProps} />
      </ArticleProvider>
    );

    // Check for search-related classes
    const hasSearchClasses =
      container.innerHTML.includes("input") ||
      container.innerHTML.includes("border") ||
      container.innerHTML.includes("rounded") ||
      container.innerHTML.includes("px-") ||
      container.innerHTML.includes("py-");

    expect(hasSearchClasses).toBe(true);
  });
});
