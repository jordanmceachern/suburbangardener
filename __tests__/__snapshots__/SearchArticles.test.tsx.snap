// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SearchArticles Component should match DOM snapshot 1`] = `
<div
  class="test-search-class"
>
  <h3
    class="text-lg font-semibold text-slate-800 dark:text-slate-200 mb-4"
  >
    Search Articles
  </h3>
  <div
    class="space-y-4"
  >
    <div
      class="relative"
    >
      <input
        class="block w-full pl-3 pr-12 py-2.5 h-10 text-base border border-slate-300 dark:border-slate-600 rounded-md bg-white dark:bg-slate-700 text-slate-900 dark:text-slate-100 placeholder-slate-900 dark:placeholder-slate-100 focus:outline-none focus:ring-2 focus:ring-primary-500 focus:border-primary-500 cursor-text hover:border-slate-400 dark:hover:border-slate-500 transition-colors"
        placeholder="Search articles..."
        type="text"
        value=""
      />
      <button
        class="absolute inset-y-0 right-0 pr-3 flex items-center text-slate-900 dark:text-slate-100 hover:text-slate-600 dark:hover:text-slate-300 disabled:opacity-50 disabled:cursor-not-allowed"
      >
        <svg
          class="h-4 w-4"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
    <div
      class="relative"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Select category"
        class="w-full px-3 py-2.5 h-10 text-base border border-slate-300 dark:border-slate-600 rounded-md bg-white dark:bg-slate-700 text-slate-900 dark:text-slate-100 focus:outline-none focus:ring-2 focus:ring-primary-500 focus:border-primary-500 cursor-pointer hover:border-slate-400 dark:hover:border-slate-500 transition-colors text-left flex items-center justify-between"
        type="button"
      >
        <span
          class="truncate"
        >
          All Categories
        </span>
        <svg
          class="h-4 w-4 text-slate-900 dark:text-slate-100 transition-transform "
          fill="none"
          stroke="currentColor"
          viewBox="0 0 20 20"
        >
          <path
            d="M6 8l4 4 4-4"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </button>
    </div>
    <div
      class="text-xs text-slate-600 dark:text-slate-300"
    >
      <span>
        0
         total articles
      </span>
    </div>
  </div>
</div>
`;
